feat(core): add pooling model initial support for V1 engine

pei0033 opened the pull request:
### 🚀 Summary of Changes

- Implemented `_pool()` method in `RBLNModelRunner` for pooling model inference based on  `GPUModelRunner` implementation
- Adjusted warmup logic to handle pooling model initialization
- Added comprehensive example scripts for Qwen3 embedding and reranker models

---

### 📌 Related Issues / Tickets

<!--
All pull requests must be linked to a Development-related Issue.
Use "Resolves/Fixes/Closes/Related to #<issue_number>" to auto-link or close the issue when merged.
-->

* Resolves #147  
---

### ✅ Type of Change

<!-- Mark all that apply using [x]. -->

* [x] ✨ Feature (`feature`)
* [ ] 🧠 Model support (`model`)
* [x] 🧬 Core engine changes (`core`)
* [ ] 🛠 Bug fix (`bug-fix`)
* [ ] ⚙️ Performance improvement (`perf`)
* [ ] 🔁 Refactor or code cleanup (`refactor`)
* [ ] 📄 Documentation (`docs`)
* [ ] ❓ Other (`other`): please describe

---

### 🧪 How to Test

**For Qwen3 Embedding**
1. Run the embedding example: `RBLN_PROFILER=0 RBLN_KERNEL_MODE=triton VLLM_RBLN_USE_VLLM_MODEL=1 VLLM_USE_V1=1 python examples/experimental/qwen3_embedding.py`
2. Verify that embedding vectors are generated for input texts
3. Expected output: Similarity scores between queries and documents.

**For Qwen3 Reranker**
1. Run the reranker example: `RBLN_PROFILER=0 RBLN_KERNEL_MODE=triton VLLM_RBLN_USE_VLLM_MODEL=1 VLLM_USE_V1=1 python examples/experimental/qwen3_reranker.py`
2. Verify that relevance scores are computed for query-document pairs
3. Expected output: Score list showing document relevance.


---

### 📋 Checklist

<!--
The PR will only be reviewed and considered for merge if the following are satisfied.
-->

* [x] PR title follows Conventional Commits format
* [x] This PR is linked to an existing issue
* [x] The test method is described, and the expected result is clearly stated
* [ ] Relevant documentation has been updated (if applicable)

---

### 💬 Notes

- The implementation follows the same pattern as upstream vLLM's V1 engine pooling model support
- Warmup logic has been adjusted to properly initialize pooling models
- This is an initial implementation; additional model types (e.g., BERT-based models) may be added in future PRs.
- The pooling function currently **runs on CPU without RBLN compilation**. Future optimization may include compiling this operation for better performance.

rebel-jiwoopark commented on vllm_rbln/v1/worker/rbln_model_runner.py:
```diff
@@ -1174,6 +1175,45 @@ def get_dp_padding(self,
+        num_scheduled_tokens: int,
+        num_scheduled_tokens_np: np.ndarray,
+        kv_connector_output: Optional[KVConnectorOutput],
+    ) -> ModelRunnerOutput:
```
I’m curious whether any modifications were made from the GPU's `_pool()`.

rebel-jiwoopark replied:
If there are any changes, I’d appreciate it if you could leave a comment on the corresponding code snippets.

rebel-jiwoopark commented:
Can this implementation also cover pooling models other than Qwen3? (such as BERT, ...)

pei0033 commented:
> Can this implementation also cover pooling models other than Qwen3? (such as BERT, ...)


I tested with other models such as BERT (`BAAI/bge-base-en-v1.5`) and Robert (`sentence-transformers/all-roberta-large-v1`), but encountered the error below when launching the engine.
 It appears that additional implementation is needed to support encoder-based models.

<details> <summary> Error log </summary>

```log
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718] EngineCore failed to start.
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718] Traceback (most recent call last):
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 709, in run_engine_core
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]     engine_core = EngineCoreProc(*args, **kwargs)
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 505, in __init__
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]     super().__init__(vllm_config, executor_class, log_stats,
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 91, in __init__
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]     self._initialize_kv_caches(vllm_config)
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 192, in _initialize_kv_caches
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]     kv_cache_configs = [
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 193, in <listcomp>
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]     get_kv_cache_config(vllm_config, kv_cache_spec_one_worker,
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/core/kv_cache_utils.py", line 1112, in get_kv_cache_config
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]     unify_hybrid_kv_cache_specs(kv_cache_spec)
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/core/kv_cache_utils.py", line 1051, in unify_hybrid_kv_cache_specs
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]     if is_kv_cache_type_uniform(kv_cache_spec):
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/core/kv_cache_utils.py", line 762, in is_kv_cache_type_uniform
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718]     _ = kv_cache_spec_values[0].merge(kv_cache_spec_values)
(EngineCore_DP0 pid=373150) [vllm-rbln] ERROR 12-03 03:43:54 core.py:718] IndexError: list index out of range
(EngineCore_DP0 pid=373150) Process EngineCore_DP0:
(EngineCore_DP0 pid=373150) Traceback (most recent call last):
(EngineCore_DP0 pid=373150)   File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
(EngineCore_DP0 pid=373150)     self.run()
(EngineCore_DP0 pid=373150)   File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
(EngineCore_DP0 pid=373150)     self._target(*self._args, **self._kwargs)
(EngineCore_DP0 pid=373150)   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 722, in run_engine_core
(EngineCore_DP0 pid=373150)     raise e
(EngineCore_DP0 pid=373150)   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 709, in run_engine_core
(EngineCore_DP0 pid=373150)     engine_core = EngineCoreProc(*args, **kwargs)
(EngineCore_DP0 pid=373150)   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 505, in __init__
(EngineCore_DP0 pid=373150)     super().__init__(vllm_config, executor_class, log_stats,
(EngineCore_DP0 pid=373150)   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 91, in __init__
(EngineCore_DP0 pid=373150)     self._initialize_kv_caches(vllm_config)
(EngineCore_DP0 pid=373150)   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 192, in _initialize_kv_caches
(EngineCore_DP0 pid=373150)     kv_cache_configs = [
(EngineCore_DP0 pid=373150)   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/engine/core.py", line 193, in <listcomp>
(EngineCore_DP0 pid=373150)     get_kv_cache_config(vllm_config, kv_cache_spec_one_worker,
(EngineCore_DP0 pid=373150)   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/core/kv_cache_utils.py", line 1112, in get_kv_cache_config
(EngineCore_DP0 pid=373150)     unify_hybrid_kv_cache_specs(kv_cache_spec)
(EngineCore_DP0 pid=373150)   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/core/kv_cache_utils.py", line 1051, in unify_hybrid_kv_cache_specs
(EngineCore_DP0 pid=373150)     if is_kv_cache_type_uniform(kv_cache_spec):
(EngineCore_DP0 pid=373150)   File "/usr/local/lib/python3.10/dist-packages/vllm/v1/core/kv_cache_utils.py", line 762, in is_kv_cache_type_uniform
(EngineCore_DP0 pid=373150)     _ = kv_cache_spec_values[0].merge(kv_cache_spec_values)
(EngineCore_DP0 pid=373150) IndexError: list index out of range
```
</details>

rebel-jaehwang commented:
hmm then I guess we need to change the signature of `AttentionMetadataBuilder.build` in the upstream. GDNAttentionMetadataBuilder would have the same problem in the upstream, though I'm not sure if there's pooling model with GDNA.

pei0033 commented:
The cases currently failing the tests appear to be those in the optimum branch rather than the torch compile branch.

pei0033 commented:
**Update on recent changes:**
1. `model_wrapper` Refactoring
    I've refactored the execution flow to always use `model_wrapper`, ensuring consistency.
    Previously, we bypassed `model_wrapper` and ran `self.model `directly when logit computation wasn't needed.
    Now, if `compute_logits` is skipped, the wrapper explicitly returns `None` for the logits. This unifies the behavior for cases like LoRA, SpecDec, and Pooling models.
2. Test Code Refactoring
    I addressed the infinite hang issue in the pooling model tests.
    This was resolved by using `patch_and_run` (in `test_model_coverage_single.py`) to execute the tests in isolated processes.
 
CC. @junstar92 (LoRA)
